### DISCUSSION TITLE
Anybody have automated testing guidelines?

### DISCUSSION THREAD

**Discussion by brianwisti** [django-commons/django-unicorn#295] [Q&A]
Subject pretty much covers the thought I'm having.

Currently I use pytest with pytest-playwright for browser-level testing. I wondered if anybody has suggestions or preferred approaches for testing the components at a lower level than Full Browser.

**Comment by adamghill** [comment]
I've only played with `playwright` a little bit, but that seems like a reasonable approach for end-to-end browser testing. Would it be useful to provide any hooks or documentation for testing the component itself?

**Comment by brianwisti** [comment]
They would, but I'm having trouble imagining what these hooks would look like. Unicorn's not just a view fragment, right? There's AJAX and JSON right from the get-go. 

Thinking out loud: if I wanted some pre-browser tests, I could make assertions that the models and objects I use in components translate cleanly to JSON. That seems to be the main surprise I bump into, especially around a couple m2m relations in my project. I'll have to ponder, and again I'm not sure where you could bake in a testing hook for that.

**Reply by GabeSoler** [reply]
Is there not a way of mocking Ajax /Jason messages with Django client? 

I think it would be really important to have an official way of testing the unicorn views.
